tools/shared.py

```diff
@@ -772,6 +772,32 @@ def clean_temp():
     prepare_to_clean_temp(EMSCRIPTEN_TEMP_DIR) # this global var might change later
   return EMSCRIPTEN_TEMP_DIR
 
+class WarningManager:
```

Comment:
I agree, having a warn method is a better implementation. I'll work on that and see if using a string dict can help clean up the code.